### PR TITLE
start adding some tests for API auth

### PR DESCRIPTION
**Pull request by czue**
Related to https://github.com/dimagi/commcare-hq/issues/28388 I thought it'd be good to have a test suite we can use to ensure we don't mess anything up.

We do have some tests around this functionality spread around, but they're all integration tests at the API-level and I thought it'd be better to be able to test the primitives.

This is just a start - planning on expanding it out tomorrow (but fine to review/merge as is)

### CODE REVIEW

File: corehq/apps/api/tests/test_auth.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertAuthenticationFail(LoginAuthentication(), self._get_request())
+
+    def test_login_with_auth(self):
+        self.assertAuthenticationSuccess(LoginAuthentication(), self._get_request_with_api_key())
+
+
+class LoginAndDomainAuthenticationTest(AuthenticationTestBase):
+
+    def test_login_with_domain(self):
+        self.assertAuthenticationSuccess(LoginAndDomainAuthentication(), self._get_request(domain=self.domain))
```
Comment (snopoke) [inline corehq/apps/api/tests/test_auth.py:65]: shouldn't this fail since it has no auth?

File: corehq/apps/api/tests/test_auth.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def assertAuthenticationSuccess(self, auth_instance, request):
+        self.assertTrue(auth_instance.is_authenticated(request))
+
+    def assertAuthenticationFail(self, auth_instance, request):
+        result = auth_instance.is_authenticated(request)
+        # currently auth classes return a 401/403 response in some scenarios
+        # this should likely be changed to always return False
+        # more discussion here: https://github.com/dimagi/commcare-hq/pull/28201#discussion_r461082885
+        if isinstance(result, HttpResponse):
+            self.assertTrue(result.status_code in (401, 403))
```
Comment (millerdev) [inline corehq/apps/api/tests/test_auth.py:48]: nit: possibly more helpful failure output
```suggestion
            self.assertIn(result.status_code, (401, 403))
```

**Review by millerdev — COMMENTED** [review]
Approved, pending @snopoke's comment.

File: corehq/apps/api/tests/test_auth.py (hunk 1 above)
Comment (czue) [reply corehq/apps/api/tests/test_auth.py]: eagle eyes. Thanks, it was returning a 401 `HttpResponse` which was not failing on `assertTrue`. This whole "sometimes it's true or false and sometimes it's an `HttpResponse` thing is really bugging me. I'm pretty sure it's entirely unneccessary, but just haven't gone through enough code to change it yet. Hopefully soon. :pray: 

(fixed in the latest 3 commits, which revealed some other issues with the domain checks...)

Comment (snopoke) [reply corehq/apps/api/tests/test_auth.py]: Seems like this is by design in Tastypie. Here's part of the docstring from `is_authenticated`:
```
Should return either ``True`` if allowed, ``False`` if not or an
        ``HttpResponse`` if you need something custom.
```

I guess the question is whether we should try and be more consistent and always return an HTTP response. I do think the custom messages are useful at times, particularly the one you linked.

**Review by snopoke — APPROVED** [review]

File: corehq/apps/api/tests/test_auth.py (hunk 1 above)
Comment (czue) [reply corehq/apps/api/tests/test_auth.py]: Yeah agreed, now that I get the use case I think the tradeoff between improvement in API docs is worth the additional code complexity

**Pull request merged**